cirq-core/cirq/ops/permutation_gate.py

```diff
@@ -74,23 +74,14 @@ def _has_unitary_(self):
         return True
 
     def _decompose_(self, qubits: Sequence['cirq.Qid']) -> 'cirq.OP_TREE':
-        n = len(qubits)
-        qubit_ids = [*range(n)]
-        is_sorted = False
-
-        def _swap_if_out_of_order(idx: int) -> Iterable['cirq.Operation']:
-            nonlocal is_sorted
-            if self._permutation[qubit_ids[idx]] > self._permutation[qubit_ids[idx + 1]]:
-                yield swap_gates.SWAP(qubits[idx], qubits[idx + 1])
-                qubit_ids[idx + 1], qubit_ids[idx] = qubit_ids[idx], qubit_ids[idx + 1]
-                is_sorted = False
-
-        while not is_sorted:
-            is_sorted = True
-            for i in range(0, n - 1, 2):
-                yield from _swap_if_out_of_order(i)
-            for i in range(1, n - 1, 2):
-                yield from _swap_if_out_of_order(i)
+        reversed_permutation_map = {v: i for i, v in enumerate(self.permutation)}
+
+        while reversed_permutation_map:
+            a = list(reversed_permutation_map.keys())[0]
```

Comment:
This code has time complexity O(num_qubits * num_cycles) but can be done in O(num_qubits), try

```py3
permutation = self.permutation.copy()

for i in range(len(permutation)):
   if permutation[i] = -1: continue
   cycle = [i]
   while permutation[cycle[-1]] != i:
      cycle.append(permutation[cylce[-1]])
   for j in cycle: permutation[j] = -1
   cycle.reverse() 
   # yield len(cycle) - 1 swap operations
   


```

Reply:
> list(reversed_permutation_map.keys())[0]
this line is O(n)

Reply:
you can make your implementation O(n) by changing that line to 

```py3
a = next(iter(reversed_permutation_map))
```

but even in that case the implementation would be a bit convoluted

Reply:
@migueltorrescosta `==`

Reply:
can you commit your code?

Reply:
all the CI checks pass for current commits, where is this test failing?
![image](https://github.com/quantumlib/Cirq/assets/5949112/3c38baba-e1a3-4ac7-87a8-c30947ac3541)

Reply:
The tests fail when I attempted this suggested solution, https://github.com/quantumlib/Cirq/pull/6588#discussion_r1593149835

Reply:
@NoureldinYosri I believe this is ready to go as is. Do mention it if there is something I should change.

Reply:
please commit the code that fails